drivers: gpio: Add GPIO driver for Intel SoC

udivya73 opened the pull request:
Modified existing driver Intel APL SoC to have a
generic driver for all Intel SoC.
Added support for GPIO driver for Intel Elkhart Lake.
The GPIO driver will support pin value read/write operations,
pin direction and interrupt configuration. ACPI enumeration
support and support for different GPIO communities is also
present.

jhedberg commented:
@udivya73 looks like there's something wrong with the `sample.yaml` content:
```
ERROR   - /workdir/zephyr/tests/boards/ehl_crb/gpio/sample.yaml: can't load (skipping): <SchemaError: error code 2: Schema validation failed:
--
  | - Key 'platforms' was not defined. Path: '/sample'.: Path: '/'>

```

jhedberg commented on tests/boards/ehl_crb/gpio/sample.yaml:
```diff
@@ -0,0 +1,14 @@
+sample:
+  description: GPIO sample app.
+  name: gpio
+  platforms: ehl_crb
```
I suspect the reason CI fails is that this should be called `platforms_allow` (not 100% sure though)

jhedberg commented:
@udivya73 you'll also need to fix this in one of the commit messages to make the Compliance check pass:
```
3: UC4 Line exceeds max length (74>72): "Added sample application to test GPIO driver for Intel Elkhart Lake board."
25
Error: Process completed with exit code 1.
``

nashif reviewed (changes requested):
This seems to be a copy of the apl driver, let decouple this from a specific platform and use just 1 driver for all platforms using this IP please.

jenmwms commented:
> This seems to be a copy of the apl driver

Agreed, this looks the same, mostly. At first glance, I see a difference with pad base implementation. This driver removes the `#define REG_PAD_BASE_ADDR		0x000C` and instead uses a `pad_base(dev)`, which I'm not sure about yet. I need to review this more.

jenmwms commented:
> let decouple this from a specific platform and use just 1 driver for all platforms using this IP please.

Hi @nashif  I'm curious on approach - Would you suggest to convert this PR to implement this unified driver? (replacing gpio_intel_*.c with a gpio_intel.c ? (So removing specific platform mention in var names, comments, etc?) ... Or would you prefer a separate PR to make it generic first, and add the remaining support for EHL GPIO in this PR?

jenmwms commented:
> @udivya73 looks like there's something wrong with the `sample.yaml` content:
> 
> ```
> ERROR   - /workdir/zephyr/tests/boards/ehl_crb/gpio/sample.yaml: can't load (skipping): <SchemaError: error code 2: Schema validation failed:
> --
>   | - Key 'platforms' was not defined. Path: '/sample'.: Path: '/'>
> ```

Perhaps the issue is the copyright/comments at the top of the file. I don't see this in the other .yaml files.

jenmwms commented on tests/boards/ehl_crb/gpio/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+tests:
+  boards.ehl_crb.gpio:
+    platform_allow: ehl_crb
+    tags: gpio
```
@udivya73 Thanks for including a sample! I'm seeing this is actually located in tests/boards/ - if your intent was to include a sample, please move it to samples/boards/ instead (adjusting dependencies as needed). Example: https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/boards/up_squared/gpio_counter

udivya73 replied:
Yes, my intent was to include a sample. I will move it.

jenmwms commented on tests/boards/ehl_crb/gpio/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+#Copyright(c)2021 Intel Corporation
```
I don't believe .yaml file needs copyright header...maybe that was the build error already mentioned?

udivya73 replied:
The build error has been resolved, I will remove the copyright header.

jenmwms commented on tests/boards/ehl_crb/gpio/src/main.c:
```diff
@@ -0,0 +1,120 @@
+ * @brief Hardware setup.
+ * Connect pins 13 and 17 of J1E2 header of EHL CRB board in loopback.
+ *
+ * @{
```
This looks like a left-over copypaste perhaps?

udivya73 replied:
I reused an existing sample app, this format is not needed here, will remove.

jenmwms commented on tests/boards/ehl_crb/gpio/src/main.c:
```diff
@@ -0,0 +1,120 @@
+	}
+}
+
+/**
```
What is the purpose of this empty doxygen comment? Copy-paste?

udivya73 replied:
I reused an existing sample app, this format is not needed here, will remove.

jenmwms commented on tests/boards/ehl_crb/gpio/prj.conf:
```diff
@@ -0,0 +1,5 @@
+# nothing here
```
please remove this comment line (I don't see this in other prj.conf, also it seems unnecessary and does not seem to add value)

jenmwms commented on tests/boards/ehl_crb/gpio/README.rst:
```diff
@@ -0,0 +1,19 @@
+.. _GPIO:
```
It would help to make this link name unique & descriptive. Example: https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/master/samples/boards/up_squared/gpio_counter/README.rst

jenmwms commented on tests/boards/ehl_crb/gpio/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.8.2)
```
Is there a reason this is an older version than others in the tree? I'm seeing `3.13.1` on several I spot checked just now.

udivya73 replied:
This is a typo, will update.

jenmwms commented on tests/boards/ehl_crb/gpio/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.8.2)
+
```
I think you are missing `find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})`. See https://lists.zephyrproject.org/g/announce/message/93

jenmwms commented on tests/boards/ehl_crb/gpio/src/main.c:
```diff
@@ -0,0 +1,120 @@
+
+/* Local Includes */
+#include <zephyr.h>
+#include <sys/printk.h>
```
Do you not need to include board.h and soc.h as well? Example: https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/boards/up_squared/gpio_counter/src/main.c

jenmwms commented on tests/boards/ehl_crb/gpio/src/main.c:
```diff
@@ -0,0 +1,120 @@
+ * this would trigger interrupts for the input pin.
+ *
+ * @brief Hardware setup.
+ * Connect pins 13 and 17 of J1E2 header of EHL CRB board in loopback.
```
Can you please help us understand the relationship for the pin numberings across these files? We see in prj.conf that it is `GPIO_COM_0_B pin 28 is used as output and pin 27 as input pin`, and here we see 13 and 17, then in a few lines we see macros for GPIO_* pins 23 and 4. This could be tricky for a user to set up correctly on their HW pins as a sample?

udivya73 replied:
Pin numbers are 28 and 27 and they map to pin 13 and 17 on the board. 23 and 4 is the pin number in the community to which the pin belongs to.

udivya73 replied:
Pins used will be 23 and 4. They map to pin 13 and 17 on the EHL CRB board.

jenmwms commented on tests/boards/ehl_crb/gpio/src/main.c:
```diff
@@ -0,0 +1,120 @@
+/* GPIO output pin number used for this sample */
+#define GPIO_OUT_PIN (23)
+/* GPIO input pin number used for this sample */
+#define GPIO_INT_PIN (4)
```
Nit - this INT is confusing. The comment says input, so is this a typo? Or is this INT for Interrupt?

udivya73 replied:
This was meant to denote INPUT.

jenmwms commented on tests/boards/ehl_crb/gpio/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.8.2)
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(NONE)
```
For `project( )`, please use a brief, descriptive name, such as examples: `gpio_counter`, or `fade_led`

jenmwms commented:
I got the same build error as buildkite. I made some comments, I hope it helps!
```
/workdir/zephyr/drivers/gpio/gpio_intel_ehl.c:595:13: error: Macro is deprecated [-Werror]
  595 | DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_EHL_DEV_CFG_DATA)
      |             ^~~~~~~~~~~~~~~~~~~~~
```

jenmwms commented on drivers/gpio/gpio_intel_ehl.c:
```diff
@@ -0,0 +1,595 @@
+									       \
+	DEVICE_DT_INST_DEFINE(n,					       \
+			      gpio_intel_ehl_init,			       \
+			      device_pm_control_nop,			       \
```
Found it! This is causing the build error reported in CI. See related recent fix for other usage removed for same reason https://github.com/zephyrproject-rtos/zephyr/pull/34704

udivya73 replied:
Thanks, will update.

jhedberg commented:
@udivya73 to make the changes and commit history clean, I'd suggest to try to split this into three commits:

 1. rename existing driver
 2. introduce changes necessary for EHL
 3. new sample app

Right now it looks like you've merged the first two steps into a single commit?

jhedberg commented on drivers/gpio/Kconfig.intel:
```diff
@@ -3,15 +3,24 @@
+	help
+	  Enable driver for Intel SoC GPIO
+
+config GPIO_INTEL_APL
```
Why do you need the separate `GPIO_INTEL_APL` and `GPIO_INTEL_EHL` Kconfig options once you have the unified `GPIO_INTEL` option? It seems like you could drop the platform-specific options.

udivya73 replied:
These configs are being used currently to determine what value to initialize DT_DRV_COMPAT with.

jhedberg replied:
Right, and what I'm saying is that you should only have a single DT binding and a single compatible string value.

jhedberg replied:
Btw, it seems like there are no users of the `intel,apollo-lake-gpio.yaml` binding (at least I couldn't find any). So that should probably be removed and then the `intel,apl-gpio.yaml` binding should be renamed to `intel,gpio.yaml` to match the new unified driver.

jhedberg commented on dts/bindings/gpio/intel,ehl-gpio.yaml:
```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2020 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+description: Intel Elkhart Lake GPIO node
```
Do you really need to add any new DTS yaml files now that we use a single driver? Seems to me like you should just make the existing APL GPIO yaml file generic the same way you made the driver generic.

jhedberg commented on samples/boards/ehl_crb/gpio/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: Example of using GPIOs on Intel Elkhart Lake board
+tests:
+  sample.board.ehl_crb.gpio_toggle:
+    platform_allow: ehl_crb
```
Would it make sense to add `up_squared` to this list now that the same driver is shared by APL and EHL?

udivya73 replied:
In that case, would we need to change the location of the app as well since it will be no longer specific to one board??

jhedberg commented on samples/boards/ehl_crb/gpio/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+sample:
+  description: Example of using GPIOs on Intel Elkhart Lake board
+  name: Example of using GPIOs on Intel Elkhart Lake board
```
Would it make sense to change these to be platform agnostic (e.g. just say "Intel boards")?

udivya73 replied:
In that case, would we need to change the location of the app as well since it will be no longer specific to one board??

jhedberg replied:
I suppose so, but I really don't want to make this a blocker for this PR, so I'm fine keeping this EHL-specific for now.

jhedberg commented:
@udivya73 it seems something didn't go quite right with your split into three commits. I would expect "
drivers: gpio: Add GPIO driver for Intel Elkhart Lake" to contain changes to `gpio_intel.c`, but right now it doesn't touch it at all. Instead it seems like those changes went into the very first commit (the one doing the renaming). Other than that this is starting to look pretty good for merging (although it would be good to test this on up_squared for any regressions first)

udivya73 committed:
drivers: gpio: Add GPIO driver for Intel Elkhart Lake

Added support for GPIO driver for Intel Elkhart Lake
board.
The GPIO driver will support pin value read/write operations,
pin direction and interrupt configuration. ACPI enumeration
support and support for different GPIO communities is also
present.

Verified on ehl_crb.

Signed-off-by: U Divya <u.divya@intel.com>

jhedberg commented on drivers/gpio/gpio_intel.c:
```diff
@@ -435,19 +440,19 @@ static int port_set_raw(const struct device *dev, uint32_t mask,
 		pin = find_lsb_set(mask) - 1;
 
-		if (pin > cfg->num_pins) {
+		if (pin >= cfg->num_pins) {
```
Can you comment on this change? Was it a bug in the APL driver, or did you update the `ngpios` property in the APL DTS so that this still does the same thing on APL?

jhedberg replied:
...and if it was a bug, it should probably be split off into its own commit, since it's not about adding EHL support per-se.

udivya73 replied:
It was a bug.  I will split off the commit, would you suggest adding it on a different PR?

jhedberg replied:
I think you can keep it in this same PR, since it all relates to the same driver

galak reviewed (dismissed):
Do we really need a board specific sample for GPIO?

jhedberg commented:
> Do we really need a board specific sample for GPIO?

@galak we don't. It has now been removed from the PR.

galak reviewed (commented):
What is dts property `community_base` I don't see it used in the code anywhere.  Am I missing something?

udivya73 committed:
drivers: gpio: Generic GPIO driver for Intel SoC

Modified GPIO driver for Intel APL SoC to have a
generic driver for all Intel SoC.

Signed-off-by: U Divya <u.divya@intel.com>

udivya73 committed:
drivers: gpio: Fixed pin number check condition

Fixed check condition for GPIO pin number.
This check is done
before reading or writing value for the pin.

Verified on ehl_crb.

Signed-off-by: U Divya <u.divya@intel.com>

udivya73 commented:
> What is dts property `community_base` I don't see it used in the code anywhere. Am I missing something?

It was used in the earlier versions of the driver code, it is no longer needed, removed the dts property.
